### ISSUE TITLE
Regression: Projection with cast throws InvalidOperationException

### ISSUE THREAD

**Issue by prochnowc** [dotnet/efcore#11529]
If i have projection which contains a type-cast i receive exception. In EF Core 2.0 it worked fine.

```
Exception message: System.InvalidOperationException: 'An exception occurred while reading a database value for property 'TestEntity.Value'. The expected type was 'System.Nullable`1[ProjectionWithCast.TestDtoEnum]' but the actual value was of type 'ProjectionWithCast.TestEnum'.'

Stack trace:    
at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityMaterializerSource.ThrowReadValueException[TValue](Exception exception, Object value, IPropertyBase property)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.EntityMaterializerSource.TryReadValue[TValue](ValueBuffer valueBuffer, Int32 index, IPropertyBase property)
   at lambda_method(Closure , QueryContext , ValueBuffer )
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ProjectionShaper.TypedProjectionShaper`3.Shape(QueryContext queryContext, ValueBuffer valueBuffer)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at ProjectionWithCast.Program.Main(String[] args
```

### Steps to reproduce

```c#
    enum TestEnum
    {
        A,
        B,
        C
    }

    class TestEntity
    {
        public int Id { get; set; }

        public TestEnum Value { get; set; }
    }

    enum TestDtoEnum
    {
        A,
        B,
        C
    }

    class TestDto
    {
        public TestDtoEnum Value { get; set; }
    }

    class TestContext : DbContext
    {
        public DbSet<TestEntity> Entities => Set<TestEntity>();

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            base.OnConfiguring(optionsBuilder);

            optionsBuilder.UseSqlServer(
                "Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=CastProjection;Integrated Security=True;Connect Timeout=30");
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var context = new TestContext();
            context.Database.EnsureCreated();

            context.Entities.Add(new TestEntity());
            context.SaveChanges();

            context.Entities.Select(e => new TestDto {Value = (TestDtoEnum) e.Value})
                   .ToList();
        }
    }
```

### Further technical details
EF Core version: 2.1.0-preview1-final
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.6.4

**Comment by ajcvickers** [comment]
Triage: need to investigate this for 2.1 since it looks like a regression.

**Comment by maumar** [comment]
no longer reproduces in the current bits, must have been fixed as part of some other bugfix, closing

**Comment by prochnowc** [comment]
@maumar Is the fix already in 2.1.0-preview2?

**Comment by maumar** [comment]
@prochnowc yes

**Comment by prochnowc** [comment]
@maumar Thanks, works like a charm. Please could you update the milestone? 2.1.0 is misleading because it's fixed in 2.1.0-preview2.
